Release notes v0.46.0 (#3246)
grafana/k6: release notes/v0.46.0.md

k6 `v0.46` is here 🎉! This release includes:

- (_optional_) `<highlight of breaking changes>`
- `<Summary of new features>` (_one or multiple bullets_)

## Breaking changes

- `#pr`, `<small_break_1>`
- `#pr`, `<small_break_2>`

### (_optional h3_) `<big_breaking_change>` `#pr`

## New features

### Loki log output sending additional headers [#3227](https://github.com/grafana/k6/pull/3227)

k6 has been able to send logs to loki for nearly 3 years since [v0.28.0](https://github.com/grafana/k6/releases/tag/v0.28.0) but didn't support any way to authenticate.

Now it can be configured to send additional headers on every request.

This is being configured with the new `header` config option similar to `label`:

```
k6 --log-output=loki=http://example.org,header.X-My-Header=123,header.Authorization=mytoken ...
```

Will now send the header `X-My-Header` with value `123` and `Authorization` with `mytoken`.

> Thanks to @federicotdn for doign the changes

Reviewer: ```suggestion
Thanks to @federicotdn for doing the changes.
```

### Cloud Traces [#3100](https://github.com/grafana/k6/pull/3100), [#3202](https://github.com/grafana/k6/pull/3202)

> This release supports the new integration between k6 and Tempo in the cloud. Grafana Cloud k6 and Grafana Cloud Traces customers will be able to start their traces in k6 using the existing `k6/experimental/tracing` module to enrich their test run analysis page with metrics and aggregations from tracing data.

Reviewer: ```suggestion
This release supports the new integration between k6 and Tempo in the cloud. [Grafana Cloud k6](https://grafana.com/products/cloud/k6) and Grafana Cloud Tempo customers will be able to start their traces in k6 using the existing `k6/experimental/tracing` module to enrich their test run analysis page with metrics and aggregations from tracing data.
```
@Blinkuu Did you mean Grafana Cloud Tempo? If Traces is the right name, can you make a suggestion for linking the product, please?

> The new cloud traces will work "out of the box" for `k6 cloud` runs. In case of `k6 run` execution, the `K6_CLOUD_TRACES_ENABLED` environment variable has to be set to `true`.

Reviewer: ```suggestion
The new Cloud traces will work "out of the box" for `k6 cloud` runs. In case of `k6 run` execution, the `K6_CLOUD_TRACES_ENABLED` environment variable has to be set to `true`.
```

Reviewer 2: Do we have a corresponding docs PR for this? I can dig in and try to figure out where this should go. 🕵️

### `<big_feature_n>` `#pr`

_what, why, and what this means for the user_

### UX improvements and enhancements

_Format as `<number> <present_verb> <object>. <credit>`_:
- _`#999` Gives terminal output prettier printing. Thanks to `@person` for the help!_

- [#3176](https://github.com/grafana/k6/pull/3176) Adds a `js/promises` package, which enables extension developers to easily create promises that will be dispatched to the eventloop using the `New` function.
- [#3181](https://github.com/grafana/k6/pull/3181) Adds a `RunOnEventLoop` method to the `modulestest.Runtime` type, which allows extensions developers to run code on the event loop from their tests.

## Bug fixes

- [#3178](https://github.com/grafana/k6/pull/3178), [xk6-grpc#23](https://github.com/grafana/xk6-grpc/pull/23) Fixes the registration of nested protobuf messages in the gRPC module. Thanks @thiagodpf for reporting and actually fixing the bug!
- [#3194](https://github.com/grafana/k6/pull/3194) Fixes loki log output exiting before push is finished.

> - [#3231](https://github.com/grafana/k6/pull/3231) Fixes the tracing module sampling option to default to 1.0 when not set by the user.

Reviewer: ```suggestion
- [#3231](https://github.com/grafana/k6/pull/3231) Fixes the tracing module sampling option to default to `1.0` when not set by the user.
```

> - [#3163](https://github.com/grafana/k6/pull/3163) Fixes our gRPC example.

Reviewer: ```suggestion
- [#3163](https://github.com/grafana/k6/pull/3163) Fixes our [gRPC example](https://github.com/grafana/k6/blob/8fa0e6b9a8b63f430df34047e4393b281ff9ee30/examples/grpc.js).
```

k6 `v0.46` is here 🎉! This release includes:

- (_optional_) `<highlight of breaking changes>`
- `<Summary of new features>` (_one or multiple bullets_)

## Breaking changes

- `#pr`, `<small_break_1>`
- `#pr`, `<small_break_2>`

### (_optional h3_) `<big_breaking_change>` `#pr`

## New features

> ### Loki log output sending additional headers [#3227](https://github.com/grafana/k6/pull/3227)

Reviewer 2: I couldn't find a corresponding PR in the docs for this, do we need to add it in this section? https://k6.io/docs/using-k6/k6-options/reference/#log-output

Reviewer 3: Yes; my bad, I'll get on it

Author: Thank you 🙇🏻

Reviewer 3: Created: https://github.com/grafana/k6-docs/pull/1293

<details>
<summary> Expand to see an example of the new functionality.</summary>

```javascript
// init phase (1)
const params = {
  "grpcbin.test.notk6.io:9001": {
    plaintext: false,
    tls: {
      cacerts: [open("cacerts0.pem")],
      cert: open("cert0.pem"),
      key: open("key0.pem"),
    }, // password omitted to demonstrate 'optional password'
  },
  "grpcbin.test.fakek6.io:9001": {
    plaintext: false,
    tls: {
      cacerts: open("cacerts1.pem"),
      cert: open("cert1.pem"),
      key: open("key1.pem"),
      password: "cert1-passphrase",
    }, // cacerts as a 'string' to demonstrate string|string[] typing of cacerts
  },
};
const clients = {
  "grpcbin.test.notk6.io:9001": new grpc.Client(),
  "grpcbin.test.fakek6.io:9001": new grpc.Client(),
};
...

// k6 - VU code phase (3)
if (__ITER === 0) {
  clients["grpcbin.test.notk6.io:9001"]
    .connect("grpcbin.test.notk6.io:9001", params["grpcbin.test.notk6.io:9001"]);
  clients["grpcbin.test.notk6.io:9001"]
    .connect("grpcbin.test.fakek6.io:9001", params["grpcbin.test.fakek6.io:9001"]);
}
...
```
</details>

> Thanks @chrismoran-mica for contribution 🙇‍♂️.

Reviewer 2: ```suggestion
Thanks @chrismoran-mica for the contribution 🙇‍♂️.
```

The full list of related PRs: [#3104](https://github.com/grafana/k6/pull/3104), [#3108](https://github.com/grafana/k6/pull/3108), [#3120](https://github.com/grafana/k6/pull/3120), [#3125](https://github.com/grafana/k6/pull/3125), [#3162](https://github.com/grafana/k6/pull/3162), [#3169](https://github.com/grafana/k6/pull/3169), [#3182](https://github.com/grafana/k6/pull/3182), [#3186](https://github.com/grafana/k6/pull/3186), [#3187](https://github.com/grafana/k6/pull/3187), [#3193](https://github.com/grafana/k6/pull/3193), [#3195](https://github.com/grafana/k6/pull/3195), [#3206](https://github.com/grafana/k6/pull/3206), [#3226](https://github.com/grafana/k6/pull/3226), [#3157](https://github.com/grafana/k6/pull/3157), [#3172](https://github.com/grafana/k6/pull/3172).

### `<big_feature_n>` `#pr`

_what, why, and what this means for the user_

### UX improvements and enhancements

_Format as `<number> <present_verb> <object>. <credit>`_:
- _`#999` Gives terminal output prettier printing. Thanks to `@person` for the help!_

> 

Reviewer: The version rendered of this section contains unexpected space between items. I don't know what is the specific reason, maybe it is this line.

```suggestion
```

- [#3176](https://github.com/grafana/k6/pull/3176) Adds a `js/promises` package, which enables extension developers to easily create promises that will be dispatched to the eventloop using the `New` function.
- [#3181](https://github.com/grafana/k6/pull/3181) Adds a `RunOnEventLoop` method to the `modulestest.Runtime` type, which allows extensions developers to run code on the event loop from their tests.

> - [#xk6-grpc#19](https://github.com/grafana/xk6-grpc/pull/19) Writing to the closed GRPC stream no longer produce the EOF (end of file) error. Additionally, if no errors' handler were set up. The error is logged to the console.

Reviewer 2: ```suggestion
- [#xk6-grpc#19](https://github.com/grafana/xk6-grpc/pull/19) Writing to the closed gRPC stream no longer produces the EOF (end of file) error. Additionally, if no error handlers were set up, the error is logged to the console.
```

## Bug fixes

> - [#3178](https://github.com/grafana/k6/pull/3178), [xk6-grpc#23](https://github.com/grafana/xk6-grpc/pull/23) Fixes the registration of nested protobuf messages in the gRPC module. Thanks @thiagodpf for reporting and actually fixing the bug!

Reviewer 2: ```suggestion
- [#3178](https://github.com/grafana/k6/pull/3178), [xk6-grpc#23](https://github.com/grafana/xk6-grpc/pull/23) Fixes the registration of nested protobuf messages in the gRPC module. Thanks @thiagodpf for reporting and fixing the bug!
```

- [#3194](https://github.com/grafana/k6/pull/3194) Fixes loki log output exiting before push is finished.
- [#3231](https://github.com/grafana/k6/pull/3231) Fixes the tracing module sampling option to default to 1.0 when not set by the user.
- [#3163](https://github.com/grafana/k6/pull/3163) Fixes our gRPC example.
- [#3196](https://github.com/grafana/k6/pull/3196) Logs packages lint fixes, and test for loki flushing at end.

## Maintenance and internal improvements

> - [#3170](https://github.com/grafana/pull/3170) Upgraded the version of gRPC k6 demo service dependency for addressing a security vulnerability ([CVE-2023-32731]).

Reviewer 2: ```suggestion
- [#3170](https://github.com/grafana/k6/pull/3170) Upgraded the version of gRPC k6 demo service dependency to address a security vulnerability ([CVE-2023-32731]).
```

    }, // password omitted to demonstrate 'optional password'
  },
  "grpcbin.test.fakek6.io:9001": {
    plaintext: false,
    tls: {
      cacerts: open("cacerts1.pem"),
      cert: open("cert1.pem"),
      key: open("key1.pem"),
      password: "cert1-passphrase",
    }, // cacerts as a 'string' to demonstrate string|string[] typing of cacerts
  },
};
const clients = {
  "grpcbin.test.notk6.io:9001": new grpc.Client(),
  "grpcbin.test.fakek6.io:9001": new grpc.Client(),
};
...

// k6 - VU code phase (3)
if (__ITER === 0) {
  clients["grpcbin.test.notk6.io:9001"]
    .connect("grpcbin.test.notk6.io:9001", params["grpcbin.test.notk6.io:9001"]);
  clients["grpcbin.test.notk6.io:9001"]
    .connect("grpcbin.test.fakek6.io:9001", params["grpcbin.test.fakek6.io:9001"]);
}
...
```
</details>

Thanks @chrismoran-mica for contribution 🙇‍♂️.

### `<big_feature_1>` `#pr`

_what, why, and what this means for the user_

### `<big_feature_n>` `#pr`

_what, why, and what this means for the user_

### UX improvements and enhancements

_Format as `<number> <present_verb> <object>. <credit>`_:
- _`#999` Gives terminal output prettier printing. Thanks to `@person` for the help!_

> - [#3157](https://github.com/grafana/k6/pull/3157) and [#3172](https://github.com/grafana/k6/pull/3172) Add a `MaxTimeSeriesInBatch` configuration option.

Author: Sounds good 👍🏻

> - [#3157](https://github.com/grafana/k6/pull/3157) and [#3172](https://github.com/grafana/k6/pull/3172) Add a `MaxTimeSeriesInBatch` configuration option.

Reviewer 4: As agreed with @codebien will be addressed in https://github.com/grafana/k6/pull/3161

- [#3177](https://github.com/grafana/k6/pull/3177) Updates the version of golangci-lint we use to the latest version.

### Cloud output v2

// TODO @codebien

* [3104](https://github.com/grafana/k6/pull/3104) Retry and flushers pool
* [3108](https://github.com/grafana/k6/pull/3108) Flush chunks
* [3120](https://github.com/grafana/k6/pull/3120) Unlimited size for body payload
* [3125](https://github.com/grafana/k6/pull/3125) Use a static remote service url
* [3162](https://github.com/grafana/k6/pull/3162) Warn about labels that have reserved names
* [3169](https://github.com/grafana/k6/pull/3169) Compact histogram
- [#3182](https://github.com/grafana/k6/pull/3182) Fix off-by-one error for spans
- [#3186](https://github.com/grafana/k6/pull/3186) Be less noisy when handling errors
- [#3187](https://github.com/grafana/k6/pull/3187) Create batches for single bucket flush
- [#3193](https://github.com/grafana/k6/pull/3193) Downscale the batch size
- [#3195](https://github.com/grafana/k6/pull/3195) Make sure to use the overwritten config
- [#3206](https://github.com/grafana/k6/pull/3206) Implement concurrent pushes accross batches

> - [#3226](https://github.com/grafana/k6/pull/3226) don't go over maxSeriesInBatch

Reviewer: What is the expectation here? Should we include all the single bits? In https://github.com/grafana/k6/pull/3161 I just linked the main issue that contains the big parts. What is the value of reporting so many details?

Reviewer 4: I share the feelings that we should not list so many details for the end user, especially for the things that are internal and are more like refactorings.

However, from the top of my head, I see one thing that is still better to stress. I mean 

```
* [3162](https://github.com/grafana/k6/pull/3162) Warn about labels that have reserved names
```

How about mentioning that the cloud output v2 drops labels that are reserved and right now it's `test_run_id` and all Prometheus system metrics (everything that prefixed with `__`).

What do you think?

Reviewer: @olegbespalov good catch, I agree to stress on it

Author: I just have made it more clear indeed. I only listed all the issues to make sure they're reflected in the end result, and to support you with putting together your bit. Feel free to remove it as you see fit and replace it with your own content 👍🏻 🙇🏻

Last but not least, this release also includes constraints on **browser contexts** usage as now only one `browserContext` is allowed per iteration, which means that the user can create a new `browserContext`, close it, and then create it again; but can not have more than one "live" `browserContext`. Instead, [scenarios](https://k6.io/docs/using-k6/scenarios/) should be used to separate independent actions in a test ([browser#929](https://github.com/grafana/xk6-browser/pull/929), [browser#945](https://github.com/grafana/xk6-browser/pull/945)).

With all these changes, a simple browser test now looks like:
```js
import { browser } from 'k6/experimental/browser';

export const options = {
  scenarios: {
    ui: {
      executor: 'shared-iterations',
      options: {
        browser: {
          type: 'chromium', // chromium is the only supported browser type so as long as
                            // the option is set, Chromium/Google Chrome will be used
        },
      },
    },
  },
};

export default async function () {
  const page = browser.newPage();
  try {
    await page.goto('https://grafana.com')
  } finally {
    page.close();
  }
}
```

### Deprecations

> - [#3121](https://github.com/grafana/k6/issues/3121) deprecates the `loadimpact/k6` docker image. We intend to keep supporting it for at most a year from now, but we recommend switching to the `grafana/k6` image as soon as possible.

Reviewer: ```suggestion
- [#3121](https://github.com/grafana/k6/issues/3121) deprecates the `loadimpact/k6` docker image. We recommend switching to the `grafana/k6` image as soon as possible.
```

It does not match what we are reporting in the docker hub page:

>THIS IMAGE IS DEPRECATED and its support will be discontinued after Dec 31, 2023. This image has been replaced by [grafana/k6](https://hub.docker.com/r/grafana/k6).

Author: Fair enough, I used what was mentioned in the PR initially 👍🏻

k6 `v0.46` is here 🎉! This release includes:

> - (_optional_) `<highlight of breaking changes>`
> - `<Summary of new features>` (_one or multiple bullets_)
> 

Reviewer 2: ```suggestion
- xk6-browser extension version bump to `v1.0.2`, which includes breaking changes.
- Send custom headers to Loki log output
- Cloud Traces, the new integration between k6 and [Tempo](https://github.com/grafana/tempo)
- Ability to configure TLS per gRPC connection
- Cloud output v2
- And many UX improvements, bug fixes, and internal improvements
```

I tried writing up this first section here, but feel free to ignore it / make edits if it doesn't fit with how it was done in the previous release notes. 😅

## Breaking changes

> - `#pr`, `<small_break_1>`
> - `#pr`, `<small_break_2>`

Reviewer 2: ```suggestion
- `#pr`, `<small_break_1>`
- `#pr`, `<small_break_2>`
```

I guess we can remove this if there are no small breaking changes?

- any tag with a key that starts with two underscores (`__`), that is marked by Prometheus convention as reserved

At the moment this is not yet the default Cloud output for the test runs executed from local machines (`k6 run -o cloud`), but it is expected to be transparently enabled in the upcoming weeks.

The full list of related PRs: [#3104](https://github.com/grafana/k6/pull/3104), [#3108](https://github.com/grafana/k6/pull/3108), [#3120](https://github.com/grafana/k6/pull/3120), [#3125](https://github.com/grafana/k6/pull/3125), [#3162](https://github.com/grafana/k6/pull/3162), [#3169](https://github.com/grafana/k6/pull/3169), [#3182](https://github.com/grafana/k6/pull/3182), [#3186](https://github.com/grafana/k6/pull/3186), [#3187](https://github.com/grafana/k6/pull/3187), [#3193](https://github.com/grafana/k6/pull/3193), [#3195](https://github.com/grafana/k6/pull/3195), [#3206](https://github.com/grafana/k6/pull/3206), [#3226](https://github.com/grafana/k6/pull/3226), [#3157](https://github.com/grafana/k6/pull/3157), [#3172](https://github.com/grafana/k6/pull/3172).

### `<big_feature_n>` `#pr`

_what, why, and what this means for the user_

> ### UX improvements and enhancements

Reviewer 2: Because this section is an h3, it's ending up under "New features"... I'm wondering if this should be an h2 instead, so it sits next to "Bug fixes", "Maintenance and internal improvements", etc?

_Format as `<number> <present_verb> <object>. <credit>`_:
- _`#999` Gives terminal output prettier printing. Thanks to `@person` for the help!_
- [#3176](https://github.com/grafana/k6/pull/3176) Adds a `js/promises` package, which enables extension developers to easily create promises that will be dispatched to the eventloop using the `New` function.
- [#3181](https://github.com/grafana/k6/pull/3181) Adds a `RunOnEventLoop` method to the `modulestest.Runtime` type, which allows extensions developers to run code on the event loop from their tests.

> - [#xk6-grpc#19](https://github.com/grafana/xk6-grpc/pull/19) Writing to the closed gRPC stream no longer produces the EOF (end of file) error. Additionally, if no error handlers were set up, the error is logged to the console.

Reviewer 2: ```suggestion
- [xk6-grpc#19](https://github.com/grafana/xk6-grpc/pull/19) Writing to the closed gRPC stream no longer produces the EOF (end of file) error. Additionally, if no error handlers were set up, the error is logged to the console.
```

Not sure how strict we want to be here with consistency, but I guess this could be changed to something like:

Fixes EOF (end of file) error that was produced when writing to a closed gRPC stream. Additionally, if no error handlers are set up, the error is logged to the console.

## Maintenance and internal improvements

### Cloud output v2

// TODO @codebien

* [3104](https://github.com/grafana/k6/pull/3104) Retry and flushers pool
* [3108](https://github.com/grafana/k6/pull/3108) Flush chunks
* [3120](https://github.com/grafana/k6/pull/3120) Unlimited size for body payload
* [3125](https://github.com/grafana/k6/pull/3125) Use a static remote service url
* [3162](https://github.com/grafana/k6/pull/3162) Warn about labels that have reserved names
* [3169](https://github.com/grafana/k6/pull/3169) Compact histogram
- [#3182](https://github.com/grafana/k6/pull/3182) Fix off-by-one error for spans
- [#3186](https://github.com/grafana/k6/pull/3186) Be less noisy when handling errors
- [#3187](https://github.com/grafana/k6/pull/3187) Create batches for single bucket flush
- [#3193](https://github.com/grafana/k6/pull/3193) Downscale the batch size
- [#3195](https://github.com/grafana/k6/pull/3195) Make sure to use the overwritten config
- [#3206](https://github.com/grafana/k6/pull/3206) Implement concurrent pushes accross batches
- [#3226](https://github.com/grafana/k6/pull/3226) don't go over maxSeriesInBatch

> ### Miscelaneous

Reviewer 5: Tiny typo fix:

```suggestion
### Miscellaneous
```

```
k6 --log-output=loki=http://example.org,header.X-My-Header=123,header.Authorization=mytoken ...
```

Will now send the header `X-My-Header` with value `123` and `Authorization` with `mytoken`.

Thanks to @federicotdn for doign the changes

### Cloud Traces [#3100](https://github.com/grafana/k6/pull/3100), [#3202](https://github.com/grafana/k6/pull/3202)

This release supports the new integration between k6 and Tempo in the cloud. Grafana Cloud k6 and Grafana Cloud Traces customers will be able to start their traces in k6 using the existing `k6/experimental/tracing` module to enrich their test run analysis page with metrics and aggregations from tracing data.

The new cloud traces will work "out of the box" for `k6 cloud` runs. In case of `k6 run` execution, the `K6_CLOUD_TRACES_ENABLED` environment variable has to be set to `true`.

### `<big_feature_1>` `#pr`

_what, why, and what this means for the user_

### `<big_feature_n>` `#pr`

_what, why, and what this means for the user_

### UX improvements and enhancements

_Format as `<number> <present_verb> <object>. <credit>`_:
- _`#999` Gives terminal output prettier printing. Thanks to `@person` for the help!_

> - TODO: [#3157](https://github.com/grafana/pull/3157) and [#3172](https://github.com/grafana/pull/3172) Add a `MaxTimeSeriesInBatch` configuration option.

Reviewer 4: Here, and in other places, the copy/paste issues with the URL, there is k6 as the repository is missing.

Author: Good catch 👀 I think the issue is that I did not copy-paste actually :D

- [#3159](https://github.com/grafana/pull/3159) Adds per-vu TLS configuration to `grpc.Client` connect method's parameters. Thanks @chrismoran-mica 🙇‍♂️.
- [#3176](https://github.com/grafana/pull/3176) Adds a `js/promises` package, which enables extension developers to easily create promises that will be dispatched to the eventloop using the `New` function.
- [#3181](https://github.com/grafana/pull/3181) Adds a `RunOnEventLoop` method to the `modulestest.Runtime` type, which allows extensions developers to run code on the event loop from their tests.

## Bug fixes

> - [#3163](https://github.com/grafana/pull/3163) Addresses a bug in our gRPC example.
> - [#3178](https://github.com/grafana/pull/3178) Fixes the registration of nested protobuf messages in the gRPC module.
> - [#3170](https://github.com/grafana/pull/3170) Upgrades the version of gRPC k6 depends on to address a security vulnerability ([CVE-2023-32731]).

Reviewer 4: The #3163 and #3170 are rather maintenance or internal improvements than bugs. So I'd suggest moving them to that section.

The #3178 was reported & fixed by the external contributor, let's add this.

```suggestion
- [#3178](https://github.com/grafana/k6/pull/3178), [xk6-grpc#23](https://github.com/grafana/xk6-grpc/pull/23) Fixes the registration of nested protobuf messages in the gRPC module. Thanks @thiagodpf for reporting and actually fixing the bug!
```

## Maintenance and internal improvements

### Cloud output v2

// TODO @codebien

* [3104](https://github.com/grafana/k6/pull/3104) Retry and flushers pool
* [3108](https://github.com/grafana/k6/pull/3108) Flush chunks
* [3120](https://github.com/grafana/k6/pull/3120) Unlimited size for body payload
* [3125](https://github.com/grafana/k6/pull/3125) Use a static remote service url
* [3162](https://github.com/grafana/k6/pull/3162) Warn about labels that have reserved names
* [3169](https://github.com/grafana/k6/pull/3169) Compact histogram
- [#3182](https://github.com/grafana/k6/pull/3182) Fix off-by-one error for spans
- [#3186](https://github.com/grafana/k6/pull/3186) Be less noisy when handling errors
- [#3187](https://github.com/grafana/k6/pull/3187) Create batches for single bucket flush
- [#3193](https://github.com/grafana/k6/pull/3193) Downscale the batch size
- [#3195](https://github.com/grafana/k6/pull/3195) Make sure to use the overwritten config
- [#3206](https://github.com/grafana/k6/pull/3206) Implement concurrent pushes accross batches
- [#3226](https://github.com/grafana/k6/pull/3226) don't go over maxSeriesInBatch

### Miscellaneous

> - [#3211](https://github.com/grafana/k6/pull/3211) Updates the experimental gRPC module version used by k6.

Reviewer 4: As part of another suggestion and rephrasing a bit.

I believe it's important to stress that the [CVE-2023-32731] is related to the demo service, so there is no rush updating to the latest version.

```suggestion
- [#3163](https://github.com/grafana/k6/pull/3163) Fixed gRPC example.
- [#3170](https://github.com/grafana/k6/pull/3170) Upgraded the version of gRPC k6 demo service dependency for addressing a security vulnerability ([CVE-2023-32731]).
- [#3211](https://github.com/grafana/k6/pull/3211) Updates the experimental gRPC module version used by k6.
```

> - [#3210](https://github.com/grafana/k6/pull/3210) Updates the xk6-output-prometheus-remote version used by k6.

Reviewer 4: I'm going to dedicate a record for the GRPC where I also mention what brings the xk6-grpc update. 

Is there anything important for our users that we should highlight in the updating of the `xk6-output-prometheus-remote` ?

Author: cc @codebien

Reviewer: @olegbespalov It does not include any functional change from a user point of view so I think we can keep it simple.

> - [#3194](https://github.com/grafana/k6/pull/3194) Fixes loki log output exiting before push is finished.

Reviewer 4: It seems like this should be in the `Bug Fixes` section.

- [#3196](https://github.com/grafana/k6/pull/3196) Logs packages lint fixes, and test for loki flushing at end.

> - [#3231](https://github.com/grafana/k6/pull/3231) Fixes the tracing module sampling option to default to 1.0 when not set by the user.

Reviewer 4: It seems like this should be in the `Bug Fixes` section.

k6 `v0.46` is here 🎉! This release includes:

- (_optional_) `<highlight of breaking changes>`
- `<Summary of new features>` (_one or multiple bullets_)

## Breaking changes

- `#pr`, `<small_break_1>`
- `#pr`, `<small_break_2>`

### (_optional h3_) `<big_breaking_change>` `#pr`

## New features

### Loki log output sending additional headers [#3227](https://github.com/grafana/k6/pull/3227)

k6 has been able to send logs to loki for nearly 3 years since [v0.28.0](https://github.com/grafana/k6/releases/tag/v0.28.0) but didn't support any way to authenticate.

Now it can be configured to send additional headers on every request.

This is being configured with the new `header` config option similar to `label`:

```
k6 --log-output=loki=http://example.org,header.X-My-Header=123,header.Authorization=mytoken ...
```

Will now send the header `X-My-Header` with value `123` and `Authorization` with `mytoken`.

> Thanks to @federicotdn for doign the changes

Reviewer: ```suggestion
Thanks to @federicotdn for doing the changes.
```

### Cloud Traces [#3100](https://github.com/grafana/k6/pull/3100), [#3202](https://github.com/grafana/k6/pull/3202)

This release supports the new integration between k6 and Tempo in the cloud. Grafana Cloud k6 and Grafana Cloud Traces customers will be able to start their traces in k6 using the existing `k6/experimental/tracing` module to enrich their test run analysis page with metrics and aggregations from tracing data.

> The new cloud traces will work "out of the box" for `k6 cloud` runs. In case of `k6 run` execution, the `K6_CLOUD_TRACES_ENABLED` environment variable has to be set to `true`.

Reviewer: ```suggestion
The new cloud traces will work _out of the box_ for runs started from the Grafana Cloud k6 app or from the `k6 cloud script.js` command. In case of `k6 run -o cloud script.js` runs, the `K6_CLOUD_TRACES_ENABLED` environment variable has to be set to `true`.
```
@Blinkuu

- TODO: [#3157](https://github.com/grafana/pull/3157) and [#3172](https://github.com/grafana/pull/3172) Add a `MaxTimeSeriesInBatch` configuration option.
- [#3159](https://github.com/grafana/pull/3159) Adds per-vu TLS configuration to `grpc.Client` connect method's parameters. Thanks @chrismoran-mica 🙇‍♂️.
- [#3176](https://github.com/grafana/pull/3176) Adds a `js/promises` package, which enables extension developers to easily create promises that will be dispatched to the eventloop using the `New` function.
- [#3181](https://github.com/grafana/pull/3181) Adds a `RunOnEventLoop` method to the `modulestest.Runtime` type, which allows extensions developers to run code on the event loop from their tests.

## Bug fixes

- [#3163](https://github.com/grafana/pull/3163) Addresses a bug in our gRPC example.
- [#3178](https://github.com/grafana/pull/3178) Fixes the registration of nested protobuf messages in the gRPC module.
- [#3170](https://github.com/grafana/pull/3170) Upgrades the version of gRPC k6 depends on to address a security vulnerability ([CVE-2023-32731]).

## Maintenance and internal improvements

### Cloud output v2

> // TODO @codebien

Reviewer: autospot as a reminder

Reviewer: https://github.com/grafana/k6/pull/3161
